File: backend/testsengine/scoring_system.py
"""
Universal Test Scoring System

A comprehensive, scalable scoring system for ALL test types in the skills assessment platform.
Supports: Multiple Choice, Numerical, Verbal, Abstract, Spatial, Situational, Diagrammatic, and more.

Author: JobGate Career Quest
Version: 1.0.0
"""

import math
import logging
from typing import Dict, Any, Optional, Union, List
from dataclasses import dataclass
from enum import Enum
from decimal import Decimal, ROUND_HALF_UP

logger = logging.getLogger(__name__)


class QuestionType(Enum):
    """Supported question types"""
    MULTIPLE_CHOICE = "multiple_choice"
    NUMERICAL = "numerical"
    VERBAL = "verbal"
    ABSTRACT = "abstract"
    SPATIAL = "spatial"
    SITUATIONAL = "situational"
    DIAGRAMMATIC = "diagrammatic"
    TECHNICAL = "technical"
    LOGICAL = "logical"
    OPEN_ENDED = "open_ended"


@dataclass
class ScoreWeight:
    """Score weight configuration for individual questions"""
    base: int = 5
    difficulty_bonus: float = 2.0
    time_factor: float = 1.0


@dataclass
class GlobalScoringConfig:
    """Global scoring configuration for tests"""
    time_weight: float = 0.3
    difficulty_weight: float = 0.5
    accuracy_weight: float = 0.2
    
    def __post_init__(self):
        """Validate that weights sum to approximately 1.0"""
        total = self.time_weight + self.difficulty_weight + self.accuracy_weight
        if abs(total - 1.0) > 0.01:
            logger.warning(f"Scoring weights sum to {total}, not 1.0. Consider normalizing.")


@dataclass
class Question:
    """Universal question object structure"""
    id: str
    type: Union[str, QuestionType]
    question: str
    options: Optional[List[str]] = None
    correct_answer: str = ""
    difficulty: int = 1
    section: int = 1
    score_weight: Optional[ScoreWeight] = None
    category: Optional[str] = None
    metadata: Optional[Dict[str, Any]] = None
    
    def __post_init__(self):
        """Initialize default values and validate"""
        if self.score_weight is None:
            self.score_weight = ScoreWeight()
        
        if isinstance(self.type, str):
            try:
                self.type = QuestionType(self.type)
            except ValueError:
                logger.warning(f"Unknown question type: {self.type}")
                self.type = QuestionType.MULTIPLE_CHOICE
        
        # Validate difficulty range
        self.difficulty = max(1, min(5, self.difficulty))
        
        # Ensure options is a list for multiple choice questions
        if self.type == QuestionType.MULTIPLE_CHOICE and self.options is None:
            self.options = []


class TimeEfficiencyCalculator:
    """Calculates time efficiency factors based on question type and difficulty"""
    
    # Time thresholds for different question types (in seconds)
    TIME_THRESHOLDS = {
        QuestionType.MULTIPLE_CHOICE: {"excellent": 15, "good": 30, "average": 60, "slow": 120},
        QuestionType.NUMERICAL: {"excellent": 30, "good": 60, "average": 90, "slow": 150},
        QuestionType.VERBAL: {"excellent": 45, "good": 90, "average": 120, "slow": 180},
        QuestionType.ABSTRACT: {"excellent": 20, "good": 40, "average": 70, "slow": 120},
        QuestionType.SPATIAL: {"excellent": 25, "good": 50, "average": 80, "slow": 130},
        QuestionType.SITUATIONAL: {"excellent": 40, "good": 80, "average": 120, "slow": 180},
        QuestionType.DIAGRAMMATIC: {"excellent": 30, "good": 60, "average": 90, "slow": 150},
        QuestionType.TECHNICAL: {"excellent": 60, "good": 120, "average": 180, "slow": 240},
        QuestionType.LOGICAL: {"excellent": 20, "good": 40, "average": 70, "slow": 120},
        QuestionType.OPEN_ENDED: {"excellent": 120, "good": 240, "average": 360, "slow": 600},
    }
    
    # Difficulty multipliers for time adjustment
    DIFFICULTY_TIME_MULTIPLIERS = {
        1: 0.8,  # Easy questions - faster expected
        2: 1.0,  # Medium questions - standard time
        3: 1.2,  # Hard questions - more time expected
        4: 1.5,  # Very hard questions - significantly more time
        5: 2.0   # Expert questions - much more time expected
    }
    
    @classmethod
    def calculate_efficiency(cls, time_taken: float, difficulty: int, question_type: QuestionType) -> float:
        """
        Calculate time efficiency factor based on time taken, difficulty, and question type.
        
        Args:
            time_taken: Time taken to answer in seconds
            difficulty: Question difficulty level (1-5)
            question_type: Type of question
            
        Returns:
            Time efficiency factor (0.0 to 2.0)
        """
        if time_taken < 0:
            logger.warning(f"Negative time taken: {time_taken}")
            return 0.0
        
        # Get base thresholds for this question type
        thresholds = cls.TIME_THRESHOLDS.get(question_type, cls.TIME_THRESHOLDS[QuestionType.MULTIPLE_CHOICE])
        
        # Adjust thresholds based on difficulty
        difficulty_multiplier = cls.DIFFICULTY_TIME_MULTIPLIERS.get(difficulty, 1.0)
        adjusted_thresholds = {
            key: value * difficulty_multiplier 
            for key, value in thresholds.items()
        }
        
        # Calculate efficiency based on adjusted thresholds
        if time_taken <= adjusted_thresholds["excellent"]:
            return 2.0  # Maximum time bonus
        elif time_taken <= adjusted_thresholds["good"]:
            return 1.5  # Good time bonus
        elif time_taken <= adjusted_thresholds["average"]:
            return 1.0  # Standard time bonus
        elif time_taken <= adjusted_thresholds["slow"]:
            return 0.5  # Reduced time bonus
        else:
            return 0.2  # Minimal time bonus for very slow answers


class AnswerValidator:
    """Validates and normalizes answers for different question types"""
    
    @staticmethod
    def normalize_answer(answer: Union[str, int, float], question_type: QuestionType) -> str:
        """
        Normalize answer to string format for comparison.
        
        Args:
            answer: User's answer in any format
            question_type: Type of question
            
        Returns:
            Normalized answer as string
        """
        if answer is None:
            return ""
        
        # Convert to string and normalize
        normalized = str(answer).strip().lower()
        
        # Handle different question types
        if question_type == QuestionType.MULTIPLE_CHOICE:
            # For multiple choice, ensure single letter
            if len(normalized) == 1 and normalized.isalpha():
                return normalized
            elif normalized in ['a', 'b', 'c', 'd', 'e']:
                return normalized
        elif question_type == QuestionType.NUMERICAL:
            # For numerical, try to normalize to decimal format
            try:
                # Remove common formatting
                cleaned = normalized.replace(',', '').replace('$', '').replace('%', '')
                # Convert to float and back to string to normalize
                num_value = float(cleaned)
                return f"{num_value:.2f}"
            except ValueError:
                return normalized
        elif question_type in [QuestionType.VERBAL, QuestionType.SITUATIONAL]:
            # For text-based questions, return as-is but normalized
            return normalized
        
        return normalized
    
    @staticmethod
    def is_answer_correct(user_answer: str, correct_answer: str, question_type: QuestionType) -> bool:
        """
        Check if user's answer is correct for the given question type.
        
        Args:
            user_answer: User's normalized answer
            correct_answer: Correct answer
            question_type: Type of question
            
        Returns:
            True if answer is correct, False otherwise
        """
        user_normalized = AnswerValidator.normalize_answer(user_answer, question_type)
        correct_normalized = AnswerValidator.normalize_answer(correct_answer, question_type)
        
        # Exact match
        if user_normalized == correct_normalized:
            return True
        
        # For numerical questions, allow some tolerance
        if question_type == QuestionType.NUMERICAL:
            try:
                user_num = float(user_normalized)
                correct_num = float(correct_normalized)
                # Allow 1% tolerance for numerical answers
                tolerance = abs(correct_num * 0.01)
                return abs(user_num - correct_num) <= tolerance
            except ValueError:
                pass
        
        # For multiple choice, handle case variations
        if question_type == QuestionType.MULTIPLE_CHOICE:
            return user_normalized == correct_normalized
        
        return False


class UniversalScoringSystem:
    """
    Universal scoring system that works for all test types.
    
    This system calculates scores based on:
    - Answer correctness (required for any points)
    - Question difficulty (higher difficulty = more points)
    - Time efficiency (faster answers = bonus points)
    - Global scoring configuration weights
    """
    
    def __init__(self, global_config: Optional[GlobalScoringConfig] = None):
        """
        Initialize the scoring system.
        
        Args:
            global_config: Global scoring configuration. If None, uses defaults.
        """
        self.global_config = global_config or GlobalScoringConfig()
        self.time_calculator = TimeEfficiencyCalculator()
        self.answer_validator = AnswerValidator()
        
        # Validate global config
        self._validate_global_config()
    
    def _validate_global_config(self):
        """Validate global scoring configuration"""
        total_weight = (self.global_config.time_weight + 
                       self.global_config.difficulty_weight + 
                       self.global_config.accuracy_weight)
        
        if abs(total_weight - 1.0) > 0.01:
            logger.warning(f"Global scoring weights sum to {total_weight}, not 1.0")
    
    def calculate_score(self, 
                       question: Question, 
                       user_answer: Union[str, int, float], 
                       time_taken: float) -> int:
        """
        Calculate score for a single question.
        
        Args:
            question: Question object with all necessary fields
            user_answer: User's answer in any format
            time_taken: Time taken to answer in seconds
            
        Returns:
            Calculated score as rounded integer
        """
        try:
            # Step 1: Check if answer is correct
            is_correct = self.answer_validator.is_answer_correct(
                user_answer, 
                question.correct_answer, 
                question.type
            )
            
            if not is_correct:
                return 0  # Wrong answer = 0 points
            
            # Step 2: Start with base score
            base_score = question.score_weight.base
            
            # Step 3: Add difficulty bonus
            difficulty_bonus = (question.difficulty * 
                              question.score_weight.difficulty_bonus * 
                              self.global_config.difficulty_weight)
            
            # Step 4: Calculate time efficiency
            time_efficiency = self.time_calculator.calculate_efficiency(
                time_taken, 
                question.difficulty, 
                question.type
            )
            
            time_bonus = time_efficiency * self.global_config.time_weight
            
            # Step 5: Calculate preliminary score
            preliminary_score = base_score + difficulty_bonus + time_bonus
            
            # Step 6: Apply accuracy weight as final multiplier
            final_score = preliminary_score * self.global_config.accuracy_weight
            
            # Step 7: Round to nearest integer
            rounded_score = int(Decimal(str(final_score)).quantize(
                Decimal('1'), rounding=ROUND_HALF_UP
            ))
            
            # Ensure non-negative score
            return max(0, rounded_score)
            
        except Exception as e:
            logger.error(f"Error calculating score for question {question.id}: {e}")
            return 0
    
    def calculate_batch_scores(self, 
                              questions: List[Question], 
                              user_answers: Dict[str, Union[str, int, float]], 
                              time_data: Dict[str, float]) -> Dict[str, int]:
        """
        Calculate scores for multiple questions at once.
        
        Args:
            questions: List of Question objects
            user_answers: Dict mapping question_id to user's answer
            time_data: Dict mapping question_id to time taken
            
        Returns:
            Dict mapping question_id to calculated score
        """
        scores = {}
        
        for question in questions:
            question_id = question.id
            user_answer = user_answers.get(question_id, "")
            time_taken = time_data.get(question_id, 0.0)
            
            score = self.calculate_score(question, user_answer, time_taken)
            scores[question_id] = score
        
        return scores
    
    def get_score_breakdown(self, 
                           question: Question, 
                           user_answer: Union[str, int, float], 
                           time_taken: float) -> Dict[str, Any]:
        """
        Get detailed breakdown of score calculation for debugging/analysis.
        
        Args:
            question: Question object
            user_answer: User's answer
            time_taken: Time taken in seconds
            
        Returns:
            Detailed breakdown of score calculation
        """
        is_correct = self.answer_validator.is_answer_correct(
            user_answer, 
            question.correct_answer, 
            question.type
        )
        
        if not is_correct:
            return {
                "correct": False,
                "final_score": 0,
                "breakdown": {
                    "base_score": question.score_weight.base,
                    "difficulty_bonus": 0,
                    "time_bonus": 0,
                    "preliminary_score": 0,
                    "accuracy_multiplier": self.global_config.accuracy_weight,
                    "final_score": 0
                }
            }
        
        base_score = question.score_weight.base
        difficulty_bonus = (question.difficulty * 
                          question.score_weight.difficulty_bonus * 
                          self.global_config.difficulty_weight)
        
        time_efficiency = self.time_calculator.calculate_efficiency(
            time_taken, 
            question.difficulty, 
            question.type
        )
        time_bonus = time_efficiency * self.global_config.time_weight
        
        preliminary_score = base_score + difficulty_bonus + time_bonus
        final_score = preliminary_score * self.global_config.accuracy_weight
        
        return {
            "correct": True,
            "final_score": int(Decimal(str(final_score)).quantize(
                Decimal('1'), rounding=ROUND_HALF_UP
            )),
            "breakdown": {
                "base_score": base_score,
                "difficulty_bonus": round(difficulty_bonus, 2),
                "time_bonus": round(time_bonus, 2),
                "time_efficiency": round(time_efficiency, 2),
                "preliminary_score": round(preliminary_score, 2),
                "accuracy_multiplier": self.global_config.accuracy_weight,
                "final_score": round(final_score, 2)
            },
            "question_info": {
                "type": question.type.value,
                "difficulty": question.difficulty,
                "time_taken": time_taken
            }
        }


class ScoringPresets:
    """Predefined scoring configurations for different test types"""
    
    @staticmethod
    def get_standard_config() -> GlobalScoringConfig:
        """Standard balanced configuration"""
        return GlobalScoringConfig(
            time_weight=0.3,
            difficulty_weight=0.5,
            accuracy_weight=0.2
        )
    
    @staticmethod
    def get_speed_focused_config() -> GlobalScoringConfig:
        """Speed-focused configuration"""
        return GlobalScoringConfig(
            time_weight=0.5,
            difficulty_weight=0.3,
            accuracy_weight=0.2
        )
    
    @staticmethod
    def get_accuracy_focused_config() -> GlobalScoringConfig:
        """Accuracy-focused configuration"""
        return GlobalScoringConfig(
            time_weight=0.2,
            difficulty_weight=0.3,
            accuracy_weight=0.5
        )
    
    @staticmethod
    def get_difficulty_focused_config() -> GlobalScoringConfig:
        """Difficulty-focused configuration"""
        return GlobalScoringConfig(
            time_weight=0.2,
            difficulty_weight=0.6,
            accuracy_weight=0.2
        )
    
    @staticmethod
    def get_numerical_config() -> GlobalScoringConfig:
        """Configuration optimized for numerical tests"""
        return GlobalScoringConfig(
            time_weight=0.4,
            difficulty_weight=0.4,
            accuracy_weight=0.2
        )
    
    @staticmethod
    def get_verbal_config() -> GlobalScoringConfig:
        """Configuration optimized for verbal tests"""
        return GlobalScoringConfig(
            time_weight=0.2,
            difficulty_weight=0.3,
            accuracy_weight=0.5
        )


# Convenience functions for easy integration
def calculate_score(question: Question, 
                   user_answer: Union[str, int, float], 
                   time_taken: float,
                   global_config: Optional[GlobalScoringConfig] = None) -> int:
    """
    Convenience function to calculate score for a single question.
    
    Args:
        question: Question object
        user_answer: User's answer
        time_taken: Time taken in seconds
        global_config: Optional global configuration
        
    Returns:
        Calculated score as integer
    """
    scoring_system = UniversalScoringSystem(global_config)
    return scoring_system.calculate_score(question, user_answer, time_taken)


def create_question(id: str, 
                   type: str, 
                   question: str, 
                   correct_answer: str,
                   difficulty: int = 1,
                   section: int = 1,
                   options: Optional[List[str]] = None,
                   score_weight: Optional[ScoreWeight] = None,
                   category: Optional[str] = None) -> Question:
    """
    Convenience function to create a Question object.
    
    Args:
        id: Unique question identifier
        type: Question type
        question: Question text
        correct_answer: Correct answer
        difficulty: Difficulty level (1-5)
        section: Section number
        options: Answer options (for multiple choice)
        score_weight: Custom score weight configuration
        category: Question category
        
    Returns:
        Question object
    """
    return Question(
        id=id,
        type=type,
        question=question,
        correct_answer=correct_answer,
        difficulty=difficulty,
        section=section,
        options=options,
        score_weight=score_weight,
        category=category
    )


# Example usage and testing
if __name__ == "__main__":
    # Example 1: Multiple Choice Question
    mc_question = create_question(
        id="q_1",
        type="multiple_choice",
        question="What is 2 + 2?",
        correct_answer="4",
        difficulty=1,
        options=["3", "4", "5", "6"]
    )
    
    scoring_system = UniversalScoringSystem()
    score = scoring_system.calculate_score(mc_question, "4", 15.0)
    print(f"Multiple Choice Score: {score}")
    
    # Example 2: Numerical Question
    num_question = create_question(
        id="q_2",
        type="numerical",
        question="If a car travels 60 mph for 2.5 hours, how far does it go?",
        correct_answer="150",
        difficulty=2
    )
    
    score = scoring_system.calculate_score(num_question, "150", 45.0)
    print(f"Numerical Score: {score}")
    
    # Example 3: Wrong Answer
    score = scoring_system.calculate_score(mc_question, "5", 10.0)
    print(f"Wrong Answer Score: {score}")
    
    # Example 4: Detailed breakdown
    breakdown = scoring_system.get_score_breakdown(num_question, "150", 45.0)
    print(f"Score Breakdown: {breakdown}")

File: frontend/src/features/skills-assessment/lib/scoringSystem.js
/**
 * Universal Test Scoring System
 *
 * A comprehensive, scalable scoring system for all test types in the skills assessment platform.
 * This system calculates scores based on accuracy, difficulty, and time efficiency.
 *
 * @author JobGate Career Quest
 * @version 1.0.0
 */

/**
 * Global scoring configuration for tests
 * @typedef {Object} GlobalScoringConfig
 * @property {number} timeWeight - Weight for time-based scoring (0-1)
 * @property {number} difficultyWeight - Weight for difficulty-based scoring (0-1)
 * @property {number} accuracyWeight - Weight for accuracy-based scoring (0-1)
 */

/**
 * Score weight configuration for individual questions
 * @typedef {Object} ScoreWeight
 * @property {number} base - Base score for correct answer
 * @property {number} difficultyBonus - Bonus multiplier for difficulty level
 * @property {number} timeFactor - Time efficiency factor multiplier
 */

/**
 * Question object structure with scoring weights
 * @typedef {Object} QuestionWithScoring
 * @property {string} id - Unique question identifier
 * @property {string} type - Question type (e.g., "multiple_choice")
 * @property {string} question - Question text
 * @property {string[]} options - Available answer options
 * @property {string} correct_answer - Correct answer identifier
 * @property {number} difficulty - Difficulty level (1-5)
 * @property {number} section - Section number
 * @property {ScoreWeight} scoreWeight - Scoring configuration
 */

/**
 * Default global scoring configuration
 * @constant {GlobalScoringConfig}
 */
export const DEFAULT_SCORING_CONFIG = {
 timeWeight: 0.3,
 difficultyWeight: 0.5,
 accuracyWeight: 0.2
};

/**
 * Default score weight configuration for questions
 * @constant {ScoreWeight}
 */
export const DEFAULT_SCORE_WEIGHT = {
 base: 5,
 difficultyBonus: 2,
 timeFactor: 1
};

/**
 * Time efficiency thresholds (in seconds)
 * @constant {Object}
 */
export const TIME_THRESHOLDS = {
 EXCELLENT: 15, // Very fast - maximum time bonus
 GOOD: 30, // Good speed - moderate time bonus
 AVERAGE: 60, // Average speed - minimal time bonus
 SLOW: 120 // Slow - no time bonus, potential penalty
};

/**
 * Difficulty level multipliers
 * @constant {Object}
 */
export const DIFFICULTY_MULTIPLIERS = {
 1: 0.8, // Easy - reduced bonus
 2: 1.0, // Medium - standard bonus
 3: 1.2, // Hard - increased bonus
 4: 1.5, // Very hard - high bonus
 5: 2.0 // Expert - maximum bonus
};

/**
 * Calculates the score for a single question based on user's answer, time taken, and scoring configuration.
 *
 * Scoring Rules:
 * 1. Wrong answer → score = 0
 * 2. Start with base score
 * 3. Add (difficulty × difficultyBonus × difficultyWeight)
 * 4. Apply time efficiency factor (faster = better, slower = smaller bonus)
 * 5. Apply accuracyWeight as final multiplier
 *
 * @param {QuestionWithScoring} question - Question object with scoring weights
 * @param {string} userAnswer - User's selected answer
 * @param {number} timeTaken - Time taken to answer (in seconds)
 * @param {GlobalScoringConfig} globalConfig - Global scoring configuration
 * @returns {number} Final score as rounded integer
 *
 * @example
 * const question = {
 * id: "logical_1_1",
 * type: "multiple_choice",
 * question: "What comes next in the series: 2, 4, 6, 8, ...?",
 * options: ["10", "12", "14", "16"],
 * correct_answer: "a",
 * difficulty: 2,
 * section: 1,
 * scoreWeight: {
 * base: 5,
 * difficultyBonus: 2,
 * timeFactor: 1
 * }
 * };
 *
 * const score = calculateScore(question, "a", 20, DEFAULT_SCORING_CONFIG);
 * console.log(score); // Returns calculated score
 */
export function calculateScore(question, userAnswer, timeTaken, globalConfig = DEFAULT_SCORING_CONFIG) {
 // Input validation
 if (!question || !userAnswer || typeof timeTaken !== 'number' || timeTaken < 0) {
 console.warn('Invalid input parameters for calculateScore');
 return 0;
 }

 // Rule 1: Wrong answer → score = 0
 if (userAnswer !== question.correct_answer) {
 return 0;
 }

 // Rule 2: Start with base score
 const baseScore = question.scoreWeight?.base || DEFAULT_SCORE_WEIGHT.base;

 // Rule 3: Add (difficulty × difficultyBonus × difficultyWeight)
 const difficulty = Math.max(1, Math.min(5, question.difficulty || 1));
 const difficultyBonus = question.scoreWeight?.difficultyBonus || DEFAULT_SCORE_WEIGHT.difficultyBonus;
 const difficultyMultiplier = DIFFICULTY_MULTIPLIERS[difficulty] || 1.0;
 const difficultyScore = difficulty * difficultyBonus * globalConfig.difficultyWeight * difficultyMultiplier;

 // Rule 4: Apply time efficiency factor
 const timeEfficiency = calculateTimeEfficiency(timeTaken, question.difficulty);
 const timeBonus = timeEfficiency * globalConfig.timeWeight;

 // Rule 5: Apply accuracyWeight as final multiplier
 const preliminaryScore = baseScore + difficultyScore + timeBonus;
 const finalScore = preliminaryScore * globalConfig.accuracyWeight;

 // Return rounded integer
 return Math.round(Math.max(0, finalScore));
}

/**
 * Calculates time efficiency factor based on time taken and question difficulty.
 * Faster answers get higher efficiency scores.
 *
 * @param {number} timeTaken - Time taken in seconds
 * @param {number} difficulty - Question difficulty level (1-5)
 * @returns {number} Time efficiency factor (0-2)
 */
export function calculateTimeEfficiency(timeTaken, difficulty = 1) {
 // Adjust thresholds based on difficulty
 const adjustedThresholds = {
 EXCELLENT: TIME_THRESHOLDS.EXCELLENT * (1 + (difficulty - 1) * 0.2),
 GOOD: TIME_THRESHOLDS.GOOD * (1 + (difficulty - 1) * 0.3),
 AVERAGE: TIME_THRESHOLDS.AVERAGE * (1 + (difficulty - 1) * 0.4),
 SLOW: TIME_THRESHOLDS.SLOW * (1 + (difficulty - 1) * 0.5)
 };

 if (timeTaken <= adjustedThresholds.EXCELLENT) {
 return 2.0; // Maximum time bonus
 } else if (timeTaken <= adjustedThresholds.GOOD) {
 return 1.5; // Good time bonus
 } else if (timeTaken <= adjustedThresholds.AVERAGE) {
 return 1.0; // Standard time bonus
 } else if (timeTaken <= adjustedThresholds.SLOW) {
 return 0.5; // Reduced time bonus
 } else {
 return 0.2; // Minimal time bonus for very slow answers
 }
}

/**
 * Calculates total test score from multiple question results.
 *
 * @param {Array} questionResults - Array of question results with scores
 * @param {GlobalScoringConfig} globalConfig - Global scoring configuration
 * @returns {Object} Total score breakdown
 */
export function calculateTotalScore(questionResults, globalConfig = DEFAULT_SCORING_CONFIG) {
 if (!Array.isArray(questionResults) || questionResults.length === 0) {
 return {
 totalScore: 0,
 maxPossibleScore: 0,
 percentage: 0,
 correctAnswers: 0,
 totalQuestions: 0,
 averageTime: 0,
 difficultyBreakdown: {}
 };
 }

 const totalScore = questionResults.reduce((sum, result) => sum + (result.score || 0), 0);
 const maxPossibleScore = questionResults.reduce((sum, result) => {
 const question = result.question;
 const baseScore = question.scoreWeight?.base || DEFAULT_SCORE_WEIGHT.base;
 const difficulty = question.difficulty || 1;
 const difficultyBonus = question.scoreWeight?.difficultyBonus || DEFAULT_SCORE_WEIGHT.difficultyBonus;
 const difficultyMultiplier = DIFFICULTY_MULTIPLIERS[difficulty] || 1.0;
 const maxDifficultyScore = difficulty * difficultyBonus * globalConfig.difficultyWeight * difficultyMultiplier;
 const maxTimeBonus = 2.0 * globalConfig.timeWeight; // Maximum time efficiency
 const maxScore = (baseScore + maxDifficultyScore + maxTimeBonus) * globalConfig.accuracyWeight;
 return sum + Math.round(maxScore);
 }, 0);

 const correctAnswers = questionResults.filter(result => result.score > 0).length;
 const totalQuestions = questionResults.length;
 const percentage = maxPossibleScore > 0 ? Math.round((totalScore / maxPossibleScore) * 100) : 0;

 const averageTime = questionResults.reduce((sum, result) => sum + (result.timeTaken || 0), 0) / totalQuestions;

 // Difficulty breakdown
 const difficultyBreakdown = {};
 questionResults.forEach(result => {
 const difficulty = result.question.difficulty || 1;
 if (!difficultyBreakdown[difficulty]) {
 difficultyBreakdown[difficulty] = { correct: 0, total: 0, totalScore: 0 };
 }
 difficultyBreakdown[difficulty].total++;
 if (result.score > 0) {
 difficultyBreakdown[difficulty].correct++;
 }
 difficultyBreakdown[difficulty].totalScore += result.score || 0;
 });

 return {
 totalScore,
 maxPossibleScore,
 percentage,
 correctAnswers,
 totalQuestions,
 averageTime: Math.round(averageTime * 100) / 100,
 difficultyBreakdown
 };
}

/**
 * Creates a question object with default scoring weights if not provided.
 *
 * @param {Object} questionData - Basic question data
 * @param {ScoreWeight} customScoreWeight - Optional custom scoring weights
 * @returns {QuestionWithScoring} Enhanced question object
 */
export function enhanceQuestionWithScoring(questionData, customScoreWeight = null) {
 return {
 ...questionData,
 scoreWeight: customScoreWeight || DEFAULT_SCORE_WEIGHT
 };
}

/**
 * Batch processes multiple questions to add scoring weights.
 *
 * @param {Array} questions - Array of question objects
 * @param {ScoreWeight} defaultScoreWeight - Default scoring weights to apply
 * @returns {Array<QuestionWithScoring>} Enhanced questions with scoring weights
 */
export function enhanceQuestionsWithScoring(questions, defaultScoreWeight = null) {
 if (!Array.isArray(questions)) {
 return [];
 }

 return questions.map(question => enhanceQuestionWithScoring(question, defaultScoreWeight));
}

/**
 * Validates scoring configuration to ensure weights sum to approximately 1.0.
 *
 * @param {GlobalScoringConfig} config - Configuration to validate
 * @returns {Object} Validation result with isValid and suggestions
 */
export function validateScoringConfig(config) {
 const sum = config.timeWeight + config.difficultyWeight + config.accuracyWeight;
 const tolerance = 0.01; // Allow small floating point differences

 return {
 isValid: Math.abs(sum - 1.0) < tolerance,
 sum,
 suggestion: sum !== 1.0 ? `Weights should sum to 1.0. Current sum: ${sum.toFixed(3)}` : null
 };
}

/**
 * Preset scoring configurations for different test types
 */
export const SCORING_PRESETS = {
 // Standard configuration - balanced approach
 STANDARD: {
 timeWeight: 0.3,
 difficultyWeight: 0.5,
 accuracyWeight: 0.2
 },

 // Speed-focused - emphasizes quick thinking
 SPEED_FOCUSED: {
 timeWeight: 0.5,
 difficultyWeight: 0.3,
 accuracyWeight: 0.2
 },

 // Accuracy-focused - emphasizes correctness over speed
 ACCURACY_FOCUSED: {
 timeWeight: 0.2,
 difficultyWeight: 0.3,
 accuracyWeight: 0.5
 },

 // Difficulty-focused - emphasizes challenging questions
 DIFFICULTY_FOCUSED: {
 timeWeight: 0.2,
 difficultyWeight: 0.6,
 accuracyWeight: 0.2
 },

 // Balanced with slight speed emphasis
 BALANCED_SPEED: {
 timeWeight: 0.4,
 difficultyWeight: 0.4,
 accuracyWeight: 0.2
 }
};

/**
 * Utility function to get a scoring preset by name
 *
 * @param {string} presetName - Name of the preset (e.g., 'STANDARD', 'SPEED_FOCUSED')
 * @returns {GlobalScoringConfig} Scoring configuration
 */
export function getScoringPreset(presetName) {
 return SCORING_PRESETS[presetName] || DEFAULT_SCORING_CONFIG;
}

/**
 * Debug utility to log detailed scoring breakdown for a question
 *
 * @param {QuestionWithScoring} question - Question object
 * @param {string} userAnswer - User's answer
 * @param {number} timeTaken - Time taken
 * @param {GlobalScoringConfig} globalConfig - Global config
 */
export function debugScoreCalculation(question, userAnswer, timeTaken, globalConfig = DEFAULT_SCORING_CONFIG) {
 console.group(`Score Calculation Debug - ${question.id}`);

 const isCorrect = userAnswer === question.correct_answer;
 console.log('Correct Answer:', isCorrect);

 if (!isCorrect) {
 console.log('Final Score: 0 (incorrect answer)');
 console.groupEnd();
 return;
 }

 const baseScore = question.scoreWeight?.base || DEFAULT_SCORE_WEIGHT.base;
 const difficulty = question.difficulty || 1;
 const difficultyBonus = question.scoreWeight?.difficultyBonus || DEFAULT_SCORE_WEIGHT.difficultyBonus;
 const difficultyMultiplier = DIFFICULTY_MULTIPLIERS[difficulty] || 1.0;
 const difficultyScore = difficulty * difficultyBonus * globalConfig.difficultyWeight * difficultyMultiplier;
 const timeEfficiency = calculateTimeEfficiency(timeTaken, difficulty);
 const timeBonus = timeEfficiency * globalConfig.timeWeight;
 const preliminaryScore = baseScore + difficultyScore + timeBonus;
 const finalScore = preliminaryScore * globalConfig.accuracyWeight;

 console.log('Base Score:', baseScore);
 console.log('Difficulty:', difficulty);
 console.log('Difficulty Multiplier:', difficultyMultiplier);
 console.log('Difficulty Score:', difficultyScore);
 console.log('Time Taken:', timeTaken, 'seconds');
 console.log('Time Efficiency:', timeEfficiency);
 console.log('Time Bonus:', timeBonus);
 console.log('Preliminary Score:', preliminaryScore);
 console.log('Accuracy Weight:', globalConfig.accuracyWeight);
 console.log('Final Score:', Math.round(finalScore));

 console.groupEnd();
}

// Export all utilities for easy importing
export default {
 calculateScore,
 calculateTimeEfficiency,
 calculateTotalScore,
 enhanceQuestionWithScoring,
 enhanceQuestionsWithScoring,
 validateScoringConfig,
 getScoringPreset,
 debugScoreCalculation,
 DEFAULT_SCORING_CONFIG,
 DEFAULT_SCORE_WEIGHT,
 TIME_THRESHOLDS,
 DIFFICULTY_MULTIPLIERS,
 SCORING_PRESETS
};

